fix(ivy): ngcc - recognize accessor members in ES5

JoostK opened the pull request:
Prior to this change, accessor functions for getters and setters would
not be considered as class member, as their declaration is vastly
different from ES2015 syntax.

With this change, the ES5 reflection host has learned to recognize the
downleveled syntax for accessors, allowing for them to be considered as
class member once again.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #28226

/cc @petebacondarwin

JoostK committed:
fix(ivy): ngcc - recognize accessor members in ES5

Prior to this change, accessor functions for getters and setters would
not be considered as class member, as their declaration is vastly
different from ES2015 syntax.

With this change, the ES5 reflection host has learned to recognize the
downleveled syntax for accessors, allowing for them to be considered as
class member once again.

Fixes #28226

gkalpak reviewed (approved):
Looks reasonable :smiley:

JoostK commented:
Pete and I had a bit of a discussion about this approach and actually identified an issue in the ES2015 host as well. We have a plan to unblock things for now, but some larger refactoring effort may be required down the line.

I'll update this PR over the weekend, presumably.

JoostK committed:
fixup! fix(ivy): ngcc - recognize accessor members in ES5

JoostK committed:
fix(ivy): ngcc - handle accessor pairs in ES2015

ngcc's reflection host needs to be able to determine all members of a
class, which it does by using the `ts.Symbol` from TypeScript's
TypeChecker.  Such Symbol however may represent multiple class members
in the case of accessors; an equally named getter/setter accessor pair
is combined into a single `ts.Symbol`.

This commit introduces logic to recognize such accessors in order for
both the getter and setter to be considered as class member, similar to
ngtsc's behavior when operating on original TypeScript code.

One difference wrt the TypeScript host is that ngcc cannot see to which
accessor originally had any decorators applied to them, as decorators
are applied to the property descriptor in general, not a specific accessor.
If an accessor has both a setter and getter, any decorators are only
attached to the setter member.

JoostK committed:
fixup! fix(ivy): ngcc - recognize accessor members in ES5

JoostK committed:
fixup! fix(ivy): ngcc - handle accessor pairs in ES2015

JoostK committed:
fixup! fix(ivy): ngcc - handle accessor pairs in ES2015

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -719,37 +722,93 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
     if (isStatic && isPropertyAccess(node)) {
       name = node.name.text;
-      value = symbol.flags & ts.SymbolFlags.Property ? node.parent.right : null;
+      value = node.parent.right;
```
👍

JoostK replied:
This was actually reverted, as it introduced a bug. Static methods would be assigned a `value` of their function expression, which failed to be statically evaluated. See 308d4f1

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts:
```diff
@@ -685,12 +686,20 @@ describe('Esm5ReflectionHost', () => {
       const getter = members.find(member => member.name === 'getter') !;
       expect(getter.kind).toEqual(ClassMemberKind.Getter);
       expect(getter.isStatic).toEqual(false);
+      expect(getter.value).toBeNull();
```
Is this `value` `null` because it is a getter or because the getter function returns `null`?

JoostK replied:
Ah, the `null` is indeed an ambiguous choice. It is `null` because it is a getter. The `implementation` member is assigned the function expression.

Come to think of it, that will probably fail to be evaluated as the static interpreter will then interpret the getter as a `Reference` to this function expression, whereas it should be the value of the member. I'm not sure how we'll actually evaluate this statically.

petebacondarwin replied:
The static evaluator will not walk into the body of an instance method (or accessor). So I don't think this matters much.

petebacondarwin reviewed (approved):
Much tidier!
